After upgrade from 4.0.1 to 5.3.1 option[:vague] raises error

mozcomp opened the pull request:
In relation to Issue: #125 

Following further tests as to cause of the issue, it relates to the interaction between the optional 3rd argument for "include seconds" and the 4th argument "options". If the 3rd argument was supplied (true/false), then the code worked as expected.
This was included and passed in spec tests.
Adding same specs, but without the optional "include_seconds" argument, gave a series of fails as found in the Issue raised.

My solution may not be acceptable, as I'm changing the arguments supplied, but by testing for a valid true/false of the "include_seconds" argument, and if "options" are nil, then swap the contents.

Once this is done, with the minor tweak of excluding the vague option before passing to the ActiveSupport support version, the code now passes all tests.

dblock commented:
I think what we really want is a test calling ActionView `time_ago_in_words(..., vague: true)`, that's the only external API. And it [is documented](https://apidock.com/rails/v5.2.3/ActionView/Helpers/DateHelper/time_ago_in_words) as something that accepts `include_seconds: true`, but nothing else. Finally, in our README we don't even show examples of `vague`. 

So... first question first, how do we want it to work?

mozcomp commented:
I may be missing something, but the issue seems to me to be whether the 3rd argument for `distance_of_time_in_words` or 2nd argument for `time_ago_in_words` /  `distance_of_time_in_words_to_now`  is optional or not.

From the name of the argument in code `include_seconds_or_options` you would think it was optional. The documentation sort of gives that impression as well.

The error isn't using `vague` in one of the helpers, it is in both, but only where `include_seconds_or_options` is used as the options hash, rather than include_seconds true/false.

There is a test for `vague` but only when used with `include_seconds: true/false`
I added the test for `vague` but without `include_seconds` - this was the failing test

So to me the question is do you make `include_seconds` mandatory and raise an `argument error` if not given `true/false`, and update the documentation accordingly, or do you push `include_seconds_or_options` to `options` if it's a hash. This is what I was doing in the PR.

mozcomp commented:
On reading your comment about `time_ago_in_words` I realised I was getting the same error as @MSCAU but mine was coming from `distance_of_time_in_words`.

My PR had already skipped a step as to the cause of the problem.

You're correct there are no tests for `time_ago_in_words(..., vague: true)` but there are for `distance_of_time_in_words(..., vague: true)` and it's at this point that we get the error.

Our entry point for the error is `helper.distance_of_time_in_words` and it's here that the code chooses to call the original method aliased as `_distance_of_time_in_words` or the new DOTIW method.

Looking at the arguments to the method `distance_of_time_in_words(from_time, to_time = 0, include_seconds_or_options = {}, options = {})` then at the DOTIW method, you'll see the expected handling of the arguments ...

```
  def distance_of_time_in_words(from_time, to_time = 0, include_seconds_or_options = {}, options = {})
    raise ArgumentError, "nil can't be converted to a Time value" if from_time.nil? || to_time.nil?
    if include_seconds_or_options.is_a?(Hash)
         options = include_seconds_or_options
     else
         options = options.dup
         options[:include_seconds] ||= !!include_seconds_or_options
      end
```  
But this is done after we have supposedly made use of the `vague` option and called the original ActionView method.

However, if the `include_seconds_or_options` argument is the `options` hash, rather than the` include_seconds` true/false, two things happen

1. the ActionView method ISN'T called because the `vague` option was in `include_seconds_or_options` not in `options`
2. the `vague` option isn't excluded from the arguments passed to the DOTIW method as expected, so is then passed to the downstream `to_sentence` where it causes throws `ArgumentError (Unknown key: :vague. Valid keys are: :words_connector, :two_words_connector, :last_word_connector, :locale)`

My thoughts are that the argument handling that is done in the DOTIW method, should be moved/duplicated to the ActionView method.

Failing that make the `include_seconds` argument mandatory thereby avoiding the whole issue.

dblock commented:
Ok I’m with you @mozcomp.

Before we do this, do you think it would be better to collapse these two arguments (`include_seconds_or_options` and `options`) into just `options` to match `time_ago_in_words`, and break backwards compatibility in a major version. I think that would fix the confusion once and for all and avoid edge cases. I’d vote to do that.

mozcomp commented:
@dblock I'm torn, the non-intrusive fix is to copy the existing argument handling within the DOTIW method to the Actionview method, this would make it work as in the original version.

However, as you say, the cleaner solution would be to collapse these arguments into a single argument `options`.

My main problem with the latter - considering how long it has been since the change that broke the gem (about 12 months I think), there is obviously only a few people implementing the gem with the 3 argument `vague` option, so if you collapse the options, we might be breaking more than we're fixing. 

I'm inclined to vote for the first option rather than the second.

dblock commented:
@mozcomp Ok, I agree we should just fix the problem in the first place

The code looks good. We also need specs for when `include_seconds_or_options = true/false` here, right?

mozcomp commented:
Ok I’ll sort the spec tomorrow - away from the office at the moment

On Mon, 21 Jun 2021 at 11:30 pm, Daniel Doubrovkine (dB.) <
***@***.***> wrote:

> @mozcomp <https://github.com/mozcomp> Ok, I agree we should just fix the
> problem in the first place
>
> The code looks good. We also need specs for when include_seconds_or_options
> = true/false here, right?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/radar/distance_of_time_in_words/pull/126#issuecomment-865056351>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AADSWVH5NYDRMZ6GEAL64MTTT5AX7ANCNFSM464XNYEA>
> .
>

mozcomp commented:
Sorry so long getting back to this ...

Added specs to cover all the permutations of the seconds & vague arguments 
Updated the readme to be more specific on both the above arguments.

Feel free to change all the above as these are all new for me to be contributing.

dblock commented:
> Sorry so long getting back to this ...
> 
> Added specs to cover all the permutations of the seconds & vague arguments
> Updated the readme to be more specific on both the above arguments.
> 
> Feel free to change all the above as these are all new for me to be contributing.

I think you forgot to push the code up here ;) Thank you!

mozcomp commented:
That's a rookie error - done now.

dblock commented:
One more thing, please add a line to https://github.com/radar/distance_of_time_in_words/blob/master/CHANGELOG.md. I think this is good to go otherwise!

dblock commented on lib/dotiw/action_view/helpers/date_helper.rb:
```diff
@@ -9,15 +9,25 @@ module DateHelper
+        if include_seconds_or_options.is_a?(Hash)
+          options = include_seconds_or_options
+        else
+          options[:include_seconds] ||= !!include_seconds_or_options
```
This modifies the incoming hash, which is bad (TM). Do `options = options.merge(...)` here.

dblock reviewed (changes requested):
One more thing I spotted.

mozcomp commented:
I think I have the arguments handling right, but I didn't like that the same code was being repeated for the two calls to the actionview methods, so moved it to a private method

dblock commented on CHANGELOG.md:
```diff
@@ -1,6 +1,6 @@
 ## 5.3.2 (Next)
 
-* Your contribution here.
```
Put this one back :)

dblock commented on CHANGELOG.md:
```diff
@@ -1,6 +1,6 @@
 ## 5.3.2 (Next)
 
-* Your contribution here.
+* [#126](https://github.com/radar/distance_of_time_in_words/pull/126): Fixes `#distance_of_time_in_words_to_now` with `vague: true` when supplied without include_seconds argument - [@mozcomp](https://github.com/mozcomp)
```
Quote `include_seconds`

dblock commented on lib/dotiw/action_view/helpers/date_helper.rb:
```diff
@@ -37,6 +29,17 @@ def distance_of_time_in_percent(from_time, current_time, to_time, options = {})
+
+      private
+      def merge_options(include_seconds_or_options, options)
+        merged_options = options.dup
```
This seems like it's doing more work than needed in all cases, duping and merging. This is simpler:

```ruby
def merge_options(include_seconds_or_options, options)
  if include_seconds_or_options.is_a?(Hash)
    options.merge(include_seconds_or_options)
  else
    options.merge(include_seconds: !!include_seconds_or_options)
  end
end
```

mozcomp replied:
Sorry, newbie error, I thought you were making the change, not waiting for me to do it.
Change as suggested

dblock reviewed (changes requested):
Some more nitpicks

dblock commented:
Thank you. See comments above, please make the minor changes to CHANGELOG, and I think this will be good to merge!

dblock commented on README.markdown:
````diff
@@ -65,7 +65,8 @@ The third argument for this method is whether or not to include seconds. By defa
 ```
 
-Yes this could just be merged into the options hash but I'm leaving it here to ensure "backwards-compatibility", because that's just an insanely radical thing to do. \m/
+Yes this could just be merged into the options hash but I'm leaving it here to ensure "backwards-compatibility", because that's just an insanely radical thing to do.  
````
I think you should put back the original author's `\m/` here.

dblock commented on CHANGELOG.md:
```diff
@@ -1,6 +1,7 @@
 
-- * Your contribution here.
+* Your contribution here.
+* [#126](https://github.com/radar/distance_of_time_in_words/pull/126): Fixes `#distance_of_time_in_words_to_now` with `vague: true` when supplied without `include_seconds` argument - [@mozcomp](https://github.com/mozcomp)
```
You're missing a period at the end. A bot will complain.

dblock commented:
Release in 5.3.2
